# Evaluated Exercise - Part 1: Databases

Please upload an html-version of your final version into the drop-zone in Moodle. If you have any issues with it, send the final version to guido.moeser@gmail.com.   
Release Date: 2025-10-07

## Final Submission Instructions

1. Complete all sections in the notebook.
2. Add explanations of all parts. **Explanations are the most important part for the grading.**
3. Comment on which configuration you found best and why. **Comments are the second most important part for the grading.**
4. Export your Jupyter Notebook to HTML and send the HTML-version.


# Topic: Building a simple Retrieval-Augmented Retrieval System with SQLite (in-memory)

### Presented by: Fidel Indalecio Mamani Maquera

## 1 – Import the required packages

**Required packages etc:**
- sqlite3
- pandas
- numpy
- from sklearn.metrics.pairwise we need cosine_similarity
- from the SentenceTransformer package we need the SentenceTransformer class


In [ ]:
# Load the minimal set of packages we will need

'''
In order to manage dependencies. in this project it will be used pixi since it is the most versatile tool.
https://pixi.sh/dev/
'''
# Uncomment below lines to installthe libraries required in this project IF using pixi package manager
# %pixi add pandas numpy torch sentence-transformers scikit-learn
import numpy as np
import pandas as pd
import sqlite3
from sentence_transformers import SentenceTransformer

/home/fidel/anaconda3/envs/jupi/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


**Explanation:**
- We will only use the core Python SQLite library and three data-science packages: pandas, numpy, and sentence-transformers.
- No extra dependencies are needed.

## 2 – Provide your own texts

Please insert 10 texts of length > 200 words (news articles, wikipedia article (parts of it), product descriptions, product reviews etc.).

In [22]:
# Each of these text samples are samples of length ~250 and are about Harry Potter's critics.
texts = [
    "The Harry Potter series is one of the most influential works of modern literature, and its impact extends far beyond fantasy fiction. J.K. Rowling’s ability to construct a detailed magical world that coexists with our own is nothing short of remarkable. From the moment readers board the Hogwarts Express, they are drawn into a setting filled with mystery, friendship, and wonder. The series’ appeal lies in its balance between escapism and emotional realism. Beneath the spells and potions, Rowling crafts a story about belonging, courage, and moral choices. Harry’s journey from an unwanted orphan to a hero who chooses compassion over revenge resonates deeply with readers of all ages. The books also celebrate education and curiosity — Hogwarts becomes a symbol of learning and self-discovery. However, what truly cements Harry Potter’s place in literary history is how it rekindled a love for reading among young audiences in a digital age. Millions of children who once disliked books found themselves captivated by this magical universe. For many, it was their first major reading adventure. While some critics question its literary sophistication, its cultural and emotional influence is undeniable. Harry Potter reminds us that imagination is a form of power — one capable of inspiring hope, unity, and resilience in a world that often feels divided.",
    "At its core, Harry Potter is not merely a fantasy series; it’s a profound coming-of-age narrative. The magic, creatures, and enchanted castles are captivating, but the emotional depth behind them is what makes the series timeless. Readers grow alongside Harry — from a lonely, mistreated child living in a cupboard to a young adult facing moral dilemmas and loss. Rowling’s genius lies in her ability to mirror the process of growing up through magic: each spell learned, each friendship tested, and each battle fought symbolizes lessons about identity and courage. The early books are playful, full of wonder and innocence, but as the series progresses, the tone darkens — reflecting both the maturation of the characters and the readers themselves. The deaths of beloved figures like Sirius and Dumbledore are not just plot points; they teach the painful but essential truth that growing up often means facing grief and uncertainty. The beauty of the story lies in its emotional honesty — magic doesn’t erase pain, but it gives people strength to face it. The final book, The Deathly Hallows, brings everything full circle, showing that real heroism comes from love, sacrifice, and self-understanding. For anyone who grew up with these books, Harry Potter is not just fiction; it’s a mirror of adolescence, courage, and transformation.",
    "The Harry Potter franchise is a cultural milestone unlike any other in recent literary history. When The Philosopher’s Stone was released in 1997, few could have predicted that it would spark a global phenomenon spanning books, films, theme parks, and countless fan communities. The series not only redefined the fantasy genre but also reshaped publishing, proving that children’s literature could captivate readers of all ages. Midnight book releases became global events; fans dressed in robes and waved wands, demonstrating the unparalleled reach of Rowling’s world-building. The story’s universal themes — love, friendship, identity, and resistance against tyranny — make it relatable across generations and cultures. Harry Potter has become more than a story; it’s a shared cultural experience. However, with such influence comes controversy. Some critics argue that the franchise’s commercialization diluted its original message, while others challenge Rowling’s personal views and their impact on the fandom. Yet, the community’s resilience — seen in fan art, fanfiction, and activism — shows how readers have made the world their own. The Harry Potter phenomenon is proof that storytelling has the power to unite people, to inspire movements, and to build lasting cultural legacies. Whether you read it as a child or discovered it later in life, the series holds a special kind of magic that transcends pages and screens.",
    "While Harry Potter is celebrated worldwide, it’s not immune to critical examination. From a literary standpoint, Rowling’s prose is often straightforward, prioritizing accessibility over stylistic complexity. Some argue that this simplicity makes the books less sophisticated compared to other fantasy epics like The Lord of the Rings or His Dark Materials. Moreover, critics point to certain limitations in world-building — particularly the underdeveloped treatment of non-European cultures and magical systems outside Britain. The series’ moral framework is also sometimes questioned. The dichotomy between “good” and “evil” can feel overly simplistic, with Slytherin House frequently vilified despite its potential for nuanced moral exploration. Additionally, Rowling’s representation of diversity has been criticized. The lack of significant non-white or LGBTQ+ characters has prompted important discussions about inclusivity in modern fantasy. Still, these criticisms do not erase the series’ achievements. Rowling succeeded in creating an emotionally compelling narrative that inspired millions to read. Her depiction of love, loss, and resilience resonates deeply with audiences. Perhaps Harry Potter’s imperfections are part of its legacy — they invite debate, reinterpretation, and reinvention by newer generations. In that sense, the series is both a triumph and a challenge, encouraging readers to imagine not only magical worlds but also more inclusive and complex ones.",
    "The Harry Potter film adaptations stand as one of the most successful cinematic projects in history. Spanning a decade and eight movies, they managed to capture the essence of Rowling’s world while offering their own artistic interpretations. The casting was near-perfect — Daniel Radcliffe, Emma Watson, and Rupert Grint embodied their roles so naturally that they became inseparable from their characters. The production design of Hogwarts, Diagon Alley, and the Ministry of Magic remains visually iconic. Yet, adapting seven dense novels into eight films was no small feat. Some fans lament missing details and subplots, like the full backstory of the Marauders or the deeper moral struggles in The Order of the Phoenix. However, the films compensated with emotional weight and cinematic beauty. The direction evolved with the story’s tone: from the whimsical innocence of The Philosopher’s Stone under Chris Columbus to the haunting realism of The Deathly Hallows directed by David Yates. The musical score by John Williams added another layer of magic, immortalizing the franchise through its unforgettable motifs. The movies not only expanded the series’ reach but also created a visual language that continues to influence fantasy filmmaking. Despite inevitable compromises, the Harry Potter films transformed literature into an enduring cinematic legend.",
    "Beyond wands and wizardry, Harry Potter is a profound meditation on morality. The series asks timeless ethical questions: What defines good and evil? Can love triumph over hate? How does power corrupt? Rowling’s narrative emphasizes that choices, not abilities, reveal a person’s true character — a philosophy echoed in Dumbledore’s wisdom. Through figures like Snape and Draco, the books explore moral ambiguity, showing that redemption and forgiveness are possible even for the deeply flawed. The theme of prejudice, especially through the treatment of “Muggle-borns” and “house-elves,” mirrors real-world discrimination. Rowling uses fantasy as a lens to discuss empathy and justice, encouraging readers to challenge bias in their own lives. Even Voldemort’s story, though steeped in evil, serves as a cautionary tale about the emptiness of power without love. The final message — that love is the most powerful magic — resonates universally. For many readers, Harry Potter functions almost like modern mythology, using the supernatural to explore what makes us human. Its moral lessons are not presented as lectures but as lived experiences, making them all the more impactful. Long after the final page, readers carry these values forward, proving that the truest magic lies not in spells, but in compassion and integrity.",
    "For millions, Harry Potter represents more than just books — it’s a cherished part of childhood. The anticipation of each new release, the smell of freshly printed pages, and the long nights reading under blankets with a flashlight are memories that shaped an entire generation. The story became a shared experience — kids and adults alike waiting at midnight for the next chapter in Harry’s life. The sense of belonging was powerful: for those who felt out of place in the real world, Hogwarts became a second home. Growing up alongside Harry, Hermione, and Ron was a journey through both fiction and adolescence. When the series ended, many readers felt they were saying goodbye to friends. Revisiting the books now evokes deep nostalgia — a reminder of simpler times when magic felt real. Even as adults, readers still find comfort in Rowling’s universe, discovering new meaning in familiar passages. The series bridges generations; parents who once read it as children now share it with their own kids. That cyclical connection makes Harry Potter timeless. More than anything, the series proves that stories can become part of who we are — shaping our imagination, our empathy, and our sense of wonder long after we leave Hogwarts behind.",
    "While Harry Potter brought joy to millions, its author’s fame has also sparked controversy. In recent years, public debate around J.K. Rowling’s personal statements has created division within the fandom. Some longtime fans feel conflicted — torn between love for the books and disappointment with their creator. This controversy raises complex questions about art and artist: can we separate the two? For many, the answer lies in the community itself. Fans have reclaimed the story, emphasizing the inclusive and accepting values they found within the pages, even when the author’s views seem to contradict them. The tension also highlights how deeply personal Harry Potter has become; it’s not just entertainment but part of people’s identities. The situation invites broader reflection on authorship and interpretation — once a story is released, does it still belong to its writer, or to the readers who give it life? Despite these challenges, the fandom remains resilient, using creativity and dialogue to sustain its magic. It’s a testament to the power of collective imagination: even when controversy clouds the legacy, the story’s message of love, courage, and unity continues to shine.",
    "Few literary fandoms rival the passion of Harry Potter fans. From early internet forums to massive conventions, the community has become a cultural force of its own. Fans don’t just consume the story — they expand it. Through fanfiction, art, podcasts, and activism, they’ve kept Hogwarts alive long after the final book. Websites like MuggleNet and The Leaky Cauldron fostered discussions, while fan-made works like A Very Potter Musical and The Wizarding World theme parks demonstrate the creativity born from this shared love. The fandom also became a space for social activism. Organizations like the Harry Potter Alliance use the story’s themes to promote equality and education, proving that fiction can inspire real-world change. This participatory culture gives Harry Potter a life independent of its author. It belongs to the readers who found empowerment and belonging within its pages. The community’s inclusivity and innovation are reminders that stories thrive when shared. The Harry Potter fandom isn’t just a byproduct of success — it’s a living example of how narrative worlds can evolve through collaboration and imagination, bridging the gap between fiction and reality.",
    "More than two decades since its debut, Harry Potter continues to enchant new generations. Its legacy extends far beyond bookshelves — influencing literature, cinema, and even education. The series taught readers to value empathy, knowledge, and courage. It reminded them that family isn’t always defined by blood and that standing up for what’s right matters, even when it’s difficult. Teachers use the books to spark discussions about ethics, prejudice, and identity. Psychologists cite them as examples of literature that nurtures emotional intelligence. Beyond academia, Harry Potter’s impact is emotional and universal. It provided solace to those who felt invisible and hope to those facing adversity. Despite controversies and criticisms, its central message remains untarnished: love conquers fear. As readers revisit the series, they discover that the true magic lies not in spells, but in the human heart. The story endures because it evolves — each generation finds new meaning in its words. Ultimately, Harry Potter is more than a cultural phenomenon; it’s a global symbol of imagination, resilience, and the timeless belief that even the smallest person can change the world."
]

## 3 – Load the SentenceTransformer model

**Tasks:**
- Load the `all-MiniLM-L6-v2` using `SentenceTransformer()` and apply it to a sentence of your choice to show the embeddings.
- Explain what the embeddings are on a short sentence.

In [ ]:
'''
    Load a lightweight embedding model from Huggingface
'''
# Use gpu to speed-up the proccess if available, otherwise use cpu.
device = 'cpu'

# The SentenceTransformer(a.k.a SBERT) is a python module that is trained on state-of-the-art embeddings models.

model = SentenceTransformer("all-MiniLM-L6-v2", device=device)

# Show example embedding of the first text from texts list
embeddings = model.encode(texts, show_progress_bar=True)
print(f'Vector column size: {embeddings[0].size}')
print(f'Vector preview: {embeddings[0]}')

Batches: 100%|██████████| 1/1 [00:01<00:00,  1.87s/it]

Vector column size: 384
Vector preview: [-4.73927520e-03 -2.87647303e-02  6.18534489e-03  5.68600297e-02
 -9.35701057e-02 -7.94375874e-03  1.64630152e-02  5.27260080e-03
  4.83171940e-02 -4.88613434e-02 -6.53464347e-02  9.64882150e-02
 -1.26995686e-02 -2.53490023e-02 -5.38840368e-02 -6.17757700e-02
  7.98225496e-03 -3.83367650e-02 -3.73081975e-02 -4.10592966e-02
  4.99279685e-02 -1.94413308e-02  3.12719867e-02 -3.30839232e-02
  1.68369282e-02 -3.10013797e-02  2.20713764e-02  3.45996977e-03
 -1.12598188e-01 -4.35473444e-03  1.29836015e-02  5.49953431e-02
 -6.88573271e-02 -8.95750970e-02 -6.97709024e-02  4.16072570e-02
  2.66878959e-02  4.41379845e-02  5.63416854e-02 -9.77397636e-02
  1.25073164e-03 -4.81840596e-02 -7.25120306e-02  6.84680641e-02
 -6.50282344e-03 -2.06968803e-02 -3.32388282e-02 -5.16091362e-02
 -4.26708050e-02  1.15191313e-02 -2.65956391e-02  1.84283387e-02
 -6.88997209e-02 -2.55718995e-02 -3.33110984e-06  1.08969081e-02
  3.79763767e-02  3.49633135e-02  6.73515946e-02 -

**Explanation:** <br>
Embeddings are numerical column vector representation of words, sentences, paragraphs, etc. In order to calculate a vector representation of the words, sentences, or paragraphs; these words, sentences, or paragraphs  have to be split into **tokens** and each token is converted to a numerical value and these numeric values are grouped together in the same order as the text and of the size of the **context window**. <br>
The context window is choosen based on many approaches(similarity, semantic context, word reelevance, etc.). Ultimately, these column vectors are called embeddings since they embed a numerical representation of the word, sentence, or paragraph.

## 4 – Create and fill an in-memory SQLite database

- Create an in-memory database with SQLite
- Create a table with the fields
  - id,
  - title,
  - text,
  - embedding,
  - import_time
- Metadata: Just add the time you loaded the data into the database
- Load the data into the database: `INSERT INTO documents (title, text, embedding, import_time) VALUES (?, ?, ?, ?)`
- Use pandas to run a SQL-request against the table to show that everything works fine

**Task:** Add the necessary SQL query. All other parts are already there.

In [ ]:
# Create an in-memory database
# in-memory means that the database is kept in CPU rathen than stored in disk.
conn = sqlite3.connect(":memory:")
cursor = conn.cursor()

# Create a simple table (embedding stored as TEXT)
# A Table called documents with fields (id, title, text, embedding, import_time) is created to store the
# texts and their corresponding embedding.
cursor.execute("""
CREATE TABLE IF NOT EXISTS documents (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    title TEXT NOT NULL,
    text TEXT NOT NULL,
    embedding BLOB,
    import_time TIMESTAMP DEFAULT CURRENT_TIMESTAMP
)
""")

# Insert documents with metadata
import datetime
# Each text in texts list is going to be stored in DB along with their corresponding embedding.
for i, t in enumerate(texts):
    title = f"Doc_{i+1}"
    emb = model.encode([t])[0]
    emb_str = ",".join(map(str, emb))   # convert vector to comma-separated string
    cursor.execute(
        "INSERT INTO documents (title, text, embedding, import_time) VALUES (?, ?, ?, ?)",
        (title, t, emb_str, str(datetime.datetime.now()))
    )

# Execute all queries to the DB so far.
conn.commit()

# Test query using pandas
pd.read_sql("SELECT id, title, import_time FROM documents", conn)

id   title                 import_time
0   1   Doc_1  2025-10-22 14:38:39.571362
1   2   Doc_2  2025-10-22 14:38:40.851272
2   3   Doc_3  2025-10-22 14:38:42.263514
3   4   Doc_4  2025-10-22 14:38:43.739052
4   5   Doc_5  2025-10-22 14:38:45.316474
5   6   Doc_6  2025-10-22 14:38:46.831493
6   7   Doc_7  2025-10-22 14:38:47.897413
7   8   Doc_8  2025-10-22 14:38:49.382484
8   9   Doc_9  2025-10-22 14:38:50.741646
9  10  Doc_10  2025-10-22 14:38:51.857552

## 5 – Define and test three similarity metrics

- **Task:** Please build to more functions. The functions should return a similarity score between two vectors.
- Will be used later to compare which metric retrieves the most relevant texts.

Here is one function:

```
# Metric 1: cosine similarity (from scikit-learn)
def cosine_sim(a, b):
    return cosine_similarity(a.reshape(1, -1), b.reshape(1, -1))[0][0]
```



In [23]:
# Metric 1: cosine similarity (from scikit-learn)
# scikit-learn uses L2 normalization. Therefore the norm will always be 1.
# utimately the cosine_sim function is equal to the dot product of the vectors.
from sklearn.metrics.pairwise import cosine_similarity
def cosine_sim(a, b):
    return cosine_similarity(a.reshape(1, -1), b.reshape(1, -1), )[0][0]

# Metric 2: dot product similarity (using linear_kernel from scikit-learn)
from sklearn.metrics.pairwise import linear_kernel
def dot_product_sim(a,b):
    return linear_kernel(a.reshape(1, -1), b.reshape(1, -1))[0][0]

# Metric 3: inverse eucleadian similarity (using euclidean_distances from scikit-learn)
from sklearn.metrics.pairwise import euclidean_distances
def inv_euclidean_sim(a,b):
    # Compute Euclidean distance
    dist = euclidean_distances(a.reshape(1, -1), b.reshape(1, -1))[0][0]
    # Convert to similarity to have a range between [0, 1]. the higher the better.
    sim = 1 / (1 + dist)
    return sim


## 6 – Build a simple Retriever

A function that encodes a query, computes the similarity between the query and each document embedding, and returns the top 3 most similar texts by default.  
  
**Task: Code is complete, please explain what happens here**

In [12]:
def retrieve(query, metric_function, top_k=3):
    # Encode the query
    q_vec = model.encode([query])[0]

    # Load all document embeddings from in-memory DB
    cursor.execute("SELECT id, title, text, embedding FROM documents")
    docs = cursor.fetchall()

    results = []
    for doc_id, title, text, emb_str in docs:
        # Embeddings in DB were stored as numeric text separeted by commas since the data can only store strings. 
        emb = np.fromstring(emb_str, sep=",")     # convert text back to numeric vector
        # Computes the similarity score between query and text iin documents.
        score = metric_function(q_vec, emb)
        # Keep the scores in a list
        results.append((title, text, score))

    # Sort by similarity and return top_k results
    results = sorted(results, key=lambda x: x[2], reverse=True)[:top_k]
    return pd.DataFrame(results, columns=["Title", "Text", "Score"])

**Explanation**<br>
The above function, <em>**retrieve**</em>, retrieves the most reelevant information (based on similarity) from the documents DB related to the query text. For this purpose it executes the next steps...<br>
- *Step 1*: Computes the embedding representation of the query text input using the exact same model used to embbed all the document texts in the DB because it is wanted to retrieve the most similar text to the query<br>
- *Step 2*: Computes the similarity score beetween each document in the DB and the query based on a *metric_function*,function that computes the similarity, then this value is kept in *results* list.<br>
- *Step 3*: Ultimately the 3 most similar scores to the query from all documents is returned.

## 7 – Run a query and inspect the results

- Run the retriever with one of the metrics (cosine_sim, dot_product_sim, or inv_euclidean_sim).
- Check how the ranking of results changes across metrics.

In [13]:
query = "Harry Potter is a franchise that almost everybody in the world has seen and loved it."
retrieve(query, cosine_sim)

Title                                               Text     Score
0   Doc_3  The Harry Potter franchise is a cultural miles...  0.679016
1   Doc_1  The Harry Potter series is one of the most inf...  0.644358
2  Doc_10  More than two decades since its debut, Harry P...  0.619767

# Fine-Tuning Phase

## 8 – Reload database with different chunk sizes and overlaps

- Reload your in-memory database with various chunk_size and overlap settings (e.g. 30/10, 60/20).
- For each configuration, insert each chunk as a new row in the documents table and repeat the insertion logic from Section 4 (using comma-separated embeddings).

**Task:** Explain what happens here and test the functions.


In [15]:
def chunk_text(text, chunk_size=50, overlap=10):
    # text is split into single words
    words = text.split()
    chunks = []
    for i in range(0, len(words), chunk_size - overlap):
        chunk = " ".join(words[i:i+chunk_size])
        if chunk.strip():
            chunks.append(chunk)
    return chunks

**Explanations:** <br>
The *chunk_text* function splits any text string into fixed subtexts(*chunk_size*) sizes, *chunk_size* is the maximum number of words that every subtext will have and the *overlap* is the number of repeated words that two consecutive subtexts will have in order to maintain the semantic context and similarity between subtexts.<br>

In [16]:
# Test the function
print(chunk_text("I really like programming since it is the best way I have to communicte my ideas with the  world.", chunk_size=7, overlap=3))

['I really like programming since it is', 'since it is the best way I', 'best way I have to communicte my', 'to communicte my ideas with the world.', 'with the world.']


## 9 – Query the new database with different similarity metrics

Compare retrieval quality under the different metrics developed above. A record which combination of metric, chunk size, and overlap yields the most meaningful matches will be printed out.

**Task:** Replace the similarity-functions here with the functions you developed above:

```
for metric in [cosine_sim, <your sim function>, <your sim function>]:
    print(f"Results using {metric.__name__}:")
    display(retrieve(query, metric))
```

In [17]:
for metric in [cosine_sim, dot_product_sim, inv_euclidean_sim]:
    print(f"Results using {metric.__name__}:")
    display(retrieve(query, metric))

Results using cosine_sim:


Title                                               Text     Score
0   Doc_3  The Harry Potter franchise is a cultural miles...  0.679016
1   Doc_1  The Harry Potter series is one of the most inf...  0.644358
2  Doc_10  More than two decades since its debut, Harry P...  0.619767

Results using dot_product_sim:


Title                                               Text     Score
0   Doc_3  The Harry Potter franchise is a cultural miles...  0.679016
1   Doc_1  The Harry Potter series is one of the most inf...  0.644358
2  Doc_10  More than two decades since its debut, Harry P...  0.619767

Results using inv_euclidean_sim:


Title                                               Text     Score
0   Doc_3  The Harry Potter franchise is a cultural miles...  0.555177
1   Doc_1  The Harry Potter series is one of the most inf...  0.542483
2  Doc_10  More than two decades since its debut, Harry P...  0.534175

## 10 – Use the systematic evaluation module

- This module systematically tests different configurations (chunk size, overlap, metric) and records which text was ranked highest.
- Visualize or summarize the outcomes to decide which configuration works best.

**Task:** Explain what happens here and run the experiment with different settings for
- chunksize
- overlap
- similarity functions

(Modify the experiment if you want, but not necessary or required).

*Please note: Replace the similarity functions with your similarity functions, otherwise it will throw an error.*

In [18]:
def evaluate_configs(query, chunk_sizes, overlaps, metrics):
    results = []

    for cs in chunk_sizes:
        for ov in overlaps:
            # Rebuild an in-memory DB for each configuration
            conn = sqlite3.connect(":memory:")
            cursor = conn.cursor()
            cursor.execute("CREATE TABLE docs (id INTEGER PRIMARY KEY, text TEXT, embedding TEXT)")

            # Insert chunks with textual embeddings
            for t in texts:
                for ch in chunk_text(t, cs, ov):
                    emb = model.encode([ch])[0]
                    emb_str = ",".join(map(str, emb))
                    cursor.execute("INSERT INTO docs (text, embedding) VALUES (?, ?)", (ch, emb_str))
            conn.commit()

            # Encode the query once
            q_vec = model.encode([query])[0]

            # Evaluate all metrics
            for metric in metrics:
                cursor.execute("SELECT id, text, embedding FROM docs")
                docs = cursor.fetchall()
                scores = []

                for _, text, emb_str in docs:
                    emb = np.fromstring(emb_str, sep=",")
                    score = metric(q_vec, emb)
                    scores.append((text, score))

                # Sort by similarity and take the best one
                top_text, top_score = sorted(scores, key=lambda x: x[1], reverse=True)[0]

                results.append((cs, ov, metric.__name__, top_text, top_score))

    # Return results as DataFrame
    return pd.DataFrame(results, columns=["ChunkSize", "Overlap", "Metric", "TopResult", "SimilarityScore"])

**Explanations:**<br>
The function, *evaluate_configs*, searches for the best configuration with the highest score based on similarity to the query and it does following the next steps.<br>
- *Step 1*: Splits the texts from the documents into subtexts(fixed length defined by *chuck_size*) and overlaps between two consecutive subtexts by the words *overlap* size in order to preserve the semantic context iin each subtext.<br>
- *Step 2*: Computes the similarity score for all the metrics input defined by user between the query input and the embeddings and by all combinations of *batch_size* options multiplied by *overlaping* options.<br>
- *Step 3*: It keeps the similarity scores in a list by preserving the most reelevant.

**Example usage 1**

In [ ]:
# Define your test query
query = "Summarize the main idea of my texts"

# Define parameter grid
chunk_sizes = [30, 50, 70]
overlaps = [5, 10, 15]
metrics = [cosine_sim, dot_product_sim, inv_euclidean_sim]

# Run evaluation
results_df = evaluate_configs(query, chunk_sizes, overlaps, metrics)

# Display full table
display(results_df)

# Optional: Find the highest similarity overall
best_config = results_df.sort_values("SimilarityScore", ascending=False).head(1)
display(best_config)


ChunkSize  Overlap             Metric  \
0          30        5         cosine_sim   
1          30        5    dot_product_sim   
2          30        5  inv_euclidean_sim   
3          30       10         cosine_sim   
4          30       10    dot_product_sim   
5          30       10  inv_euclidean_sim   
6          30       15         cosine_sim   
7          30       15    dot_product_sim   
8          30       15  inv_euclidean_sim   
9          50        5         cosine_sim   
10         50        5    dot_product_sim   
11         50        5  inv_euclidean_sim   
12         50       10         cosine_sim   
13         50       10    dot_product_sim   
14         50       10  inv_euclidean_sim   
15         50       15         cosine_sim   
16         50       15    dot_product_sim   
17         50       15  inv_euclidean_sim   
18         70        5         cosine_sim   
19         70        5    dot_product_sim   
20         70        5  inv_euclidean_sim   
21         70       10         cosine_sim   
22         70       10    dot_product_sim   
23         70       10  inv_euclidean_sim   
24         70       15         cosine_sim   
25         70       15    dot_product_sim   
26         70       15  inv_euclidean_sim   

                                            TopResult  SimilarityScore  
0   up for what’s right matters, even when it’s di...         0.352712  
1   up for what’s right matters, even when it’s di...         0.352712  
2   up for what’s right matters, even when it’s di...         0.467772  
3   human. Its moral lessons are not presented as ...         0.359960  
4   human. Its moral lessons are not presented as ...         0.359960  
5   human. Its moral lessons are not presented as ...         0.469174  
6   not presented as lectures but as lived experie...         0.328957  
7   not presented as lectures but as lived experie...         0.328957  
8   not presented as lectures but as lived experie...         0.463288  
9   to unite people, to inspire movements, and to ...         0.314964  
10  to unite people, to inspire movements, and to ...         0.314964  
11  to unite people, to inspire movements, and to ...         0.460723  
12  It’s a testament to the power of collective im...         0.307367  
13  It’s a testament to the power of collective im...         0.307367  
14  It’s a testament to the power of collective im...         0.459353  
15  that storytelling has the power to unite peopl...         0.341899  
16  that storytelling has the power to unite peopl...         0.341899  
17  that storytelling has the power to unite peopl...         0.465710  
18  Potter’s place in literary history is how it r...         0.270113  
19  Potter’s place in literary history is how it r...         0.270113  
20  Potter’s place in literary history is how it r...         0.452855  
21  to unite people, to inspire movements, and to ...         0.314964  
22  to unite people, to inspire movements, and to ...         0.314964  
23  to unite people, to inspire movements, and to ...         0.460723  
24  power of collective imagination: even when con...         0.306765  
25  power of collective imagination: even when con...         0.306765  
26  power of collective imagination: even when con...         0.459245

ChunkSize  Overlap             Metric  \
5         30       10  inv_euclidean_sim   

                                           TopResult  SimilarityScore  
5  human. Its moral lessons are not presented as ...         0.469174

**Example usage 2**

In [21]:
# Define your test query
query = '''
Harry Potter is a franchise that almost everybody in the world has seen and loved it. 
This is someting that has never seen in any other franchise before. This comment is probably to positive.
'''

# Define parameter grid
chunk_sizes = [30, 50, 90]
overlaps = [7, 15, 20]
metrics = [cosine_sim, dot_product_sim, inv_euclidean_sim]

# Run evaluation
results_df = evaluate_configs(query, chunk_sizes, overlaps, metrics)

# Display full table
display(results_df)

# Optional: Find the highest similarity overall
best_config = results_df.sort_values("SimilarityScore", ascending=False).head(1)
display(best_config)

ChunkSize  Overlap             Metric  \
0          30        7         cosine_sim   
1          30        7    dot_product_sim   
2          30        7  inv_euclidean_sim   
3          30       15         cosine_sim   
4          30       15    dot_product_sim   
5          30       15  inv_euclidean_sim   
6          30       20         cosine_sim   
7          30       20    dot_product_sim   
8          30       20  inv_euclidean_sim   
9          50        7         cosine_sim   
10         50        7    dot_product_sim   
11         50        7  inv_euclidean_sim   
12         50       15         cosine_sim   
13         50       15    dot_product_sim   
14         50       15  inv_euclidean_sim   
15         50       20         cosine_sim   
16         50       20    dot_product_sim   
17         50       20  inv_euclidean_sim   
18         90        7         cosine_sim   
19         90        7    dot_product_sim   
20         90        7  inv_euclidean_sim   
21         90       15         cosine_sim   
22         90       15    dot_product_sim   
23         90       15  inv_euclidean_sim   
24         90       20         cosine_sim   
25         90       20    dot_product_sim   
26         90       20  inv_euclidean_sim   

                                            TopResult  SimilarityScore  
0   More than two decades since its debut, Harry P...         0.691213  
1   More than two decades since its debut, Harry P...         0.691213  
2   More than two decades since its debut, Harry P...         0.559955  
3   Harry Potter has become more than a story; it’...         0.729230  
4   Harry Potter has become more than a story; it’...         0.729230  
5   Harry Potter has become more than a story; it’...         0.576072  
6   relatable across generations and cultures. Har...         0.781292  
7   relatable across generations and cultures. Har...         0.781292  
8   relatable across generations and cultures. Har...         0.601912  
9   story’s universal themes — love, friendship, i...         0.737604  
10  story’s universal themes — love, friendship, i...         0.737604  
11  story’s universal themes — love, friendship, i...         0.579903  
12  Harry Potter has become more than a story; it’...         0.724898  
13  Harry Potter has become more than a story; it’...         0.724898  
14  Harry Potter has become more than a story; it’...         0.574133  
15  love, friendship, identity, and resistance aga...         0.742812  
16  love, friendship, identity, and resistance aga...         0.742812  
17  love, friendship, identity, and resistance aga...         0.582343  
18  franchise through its unforgettable motifs. Th...         0.748713  
19  franchise through its unforgettable motifs. Th...         0.748713  
20  franchise through its unforgettable motifs. Th...         0.585163  
21  its words. Ultimately, Harry Potter is more th...         0.682321  
22  its words. Ultimately, Harry Potter is more th...         0.682321  
23  its words. Ultimately, Harry Potter is more th...         0.556454  
24  views and their impact on the fandom. Yet, the...         0.661916  
25  views and their impact on the fandom. Yet, the...         0.661916  
26  views and their impact on the fandom. Yet, the...         0.548759

ChunkSize  Overlap      Metric  \
6         30       20  cosine_sim   

                                           TopResult  SimilarityScore  
6  relatable across generations and cultures. Har...         0.781292

**Interpretation of Results**<br>
**<em>Example 1</em>** shows very low similarity (avg. ~46%) since the query is not related to Harry Potter franchise. However the **<em>Example 2</em>** shows very high(avg. ~70%) similarity score because the query is very related to a positive critic that some documents also share.<br>
Another important observation is that the metric *cosine_sim* and  *dot_product_sim* are always the same since the *cosine_similarity* from *scikit-learn* uses L2 normalization and therefore the magnitude of the vectors will always be 1. Consequently, *cosine_sim* and  *dot_product_sim* of two vectors will always provide the exact same value.<br>
The diferent metrics also shows that *cosine_similarity* is always higher than the *euclidean_similarity*. This means that *euclidean_similarity* could be used for very very similar documents and *cosine_similarity* should be used for average common texts (which is used by most approaches in the industry.)<br>
Ultimately, combining several *chunk_sizes* and *overlaps* demonstrates the fine-tune works better when the size of the query is similar to the *chunk_sizes*. However, all the texts returned are the same so probably longer texts(~2000 words per text) are needed in order to see different text results by using *chunk_sizes* and *overlaps*.